soroban-env-host/src/host/diagnostics_helper.rs

```diff
@@ -0,0 +1,105 @@
+use soroban_env_common::{
+    xdr::{ContractEventType, Hash, ScVal, ScVec},
+    ConversionError, ScValObjRef, Symbol, TryFromVal, TryIntoVal, VecObject,
+};
+
+use crate::events::{InternalContractEvent, InternalEvent};
+use crate::{budget::AsBudget, Host, HostError, RawVal};
+
+/// None of these functions are metered, which is why they're behind the is_debug check
+impl Host {
+    fn hash_to_scval(&self, hash: Hash) -> Result<ScVal, HostError> {
+        Ok(ScVal::Bytes(self.scbytes_from_hash(&hash)?))
+    }
+
+    fn record_system_debug_contract_event(
+        &self,
+        type_: ContractEventType,
+        topics: VecObject,
+        data: RawVal,
+    ) -> Result<(), HostError> {
+        let ce = InternalContractEvent {
+            type_,
+            contract_id: None,
+            topics,
+            data,
+        };
+        self.get_events_mut(|events| {
+            Ok(events.record(InternalEvent::StructuredDebug(ce), self.as_budget()))
+        })?
+    }
+
+    // Emits an event with topic = ["fn_call", contract_id, function_name] and
+    // data = [arg1, args2, ...]
+    pub fn fn_call_diagnostics(
+        &self,
+        contract_id: &Hash,
+        func: &Symbol,
+        args: &[RawVal],
+    ) -> Result<(), HostError> {
+        if !self.is_debug() {
+            return Ok(());
+        }
+
+        self.as_budget().with_free_budget(|| {
+            let mut topics: Vec<ScVal> = Vec::new();
+            topics.push(ScVal::Symbol(crate::xdr::ScSymbol(
+                self.map_err("fn_call".as_bytes().try_into())?,
+            )));
+
+            topics.push(self.hash_to_scval(contract_id.clone())?);
+            topics.push(func.try_into_val(self)?);
+
+            let scvec_topics: ScVec = self.map_err(topics.try_into())?;
+            let scval_topics = ScVal::Vec(Some(scvec_topics));
+            let topics_obj_ref = ScValObjRef::classify(&scval_topics)
```

Comment:
Good point. I updated how all of this works.

Reply:
great! One further minor nit is that if you're building an always-3-element vector, you can get away with:

```
let mut v = Vec::with_capacity(3);
v.push(a)
v.push(b)
v.push(c)
```

to avoid reallocating, or even more directly:

```
let v = vec![a, b, c]
```

which I _think_ will manage to optimize to a bit less code still.